### PR TITLE
Paper: USACE Coastal Engineering Toolkit and a Method of Creating a Web-Based Application

### PR DESCRIPTION
**Pull request by acatlett8**
If you are creating this PR in order to submit a draft of your paper,
see http://procbuild.scipy.org/ for logs generated by the build
process.

See the [project readme](https://github.com/scipy-conference/scipy_proceedings#build-your-paper)
for more information.

*Editor*: @cbcunc

*Reviewers*: @KedoKudo

### CODE REVIEW

**Comment by scoobies** [comment]
Hi!, I'm @scoobies, a friendly bot.

Type ```@scoobies help``` to discover how I can help you.

**Comment by deniederhut** [comment]
@scoobies mark paper

**Comment by deniederhut** [comment]
@scoobies build status

**Comment by scoobies** [comment]
{"erdc-paper_catlett"=>{"data"=>{"build_output"=>"[*] Updating proceedings build tools in {self.master_repo_path}...\nAlready up to date.\n[*] Check out paper repository...\nCloning into '/tmp/tmp_ut2kxjw'...\nMoving proceedings build tools to temp directory...\n[*] Build the paper...\nBuilding: Catlett_Amanda\n*** Warning: /tmp/tmp_ut2kxjw/publisher/../output/Catlett_Amanda/paper_stats.json does not exist.\n", "build_pdf_path"=>"/procbuild/procbuild/../cache/erdc-paper_catlett.pdf", "build_status"=>"success", "build_timestamp"=>"29/05 01:34"}, "status"=>"success"}}

**Comment by deniederhut** [comment]
@scoobies mark needs review

**Comment by deniederhut** [comment]
@scoobies assign @cbcunc as editor

**Comment by scoobies** [comment]
Assigned! @cbcunc is now the editor

**Comment by mepa** [comment]
@scoobies add @KedoKudo to reviewers

**Comment by scoobies** [comment]
@KedoKudo added to the reviewers list!

**Comment by cbcunc** [comment]
Editor confirms paper builds.

**Comment by deniederhut** [comment]
Hi! We've moved the deadline for first reviews back one week, to June 16th, to give us a little more time to find reviewers for all the papers this year. You can see the updated full timeline at https://github.com/scipy-conference/scipy_proceedings#timeline-for-2022.

**Comment by cbcunc** [comment]
@KedoKudo First review is due June 16.

File: papers/Catlett_Amanda/UCET.rst [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+Originally ACES was written in FORTRAN 77 and as technology has evolved, the ability to use the tool has decreased.
+In 2017, the codebase was converted from FORTRAN 77 to MATLAB and Python. This conversion ensured that coastal engineers
+using this tool base would not need training in yet another coding language. In 2020, the Engineered Resilient Systems
+(ERS) Rapid Application Development (RAD) team undertook the project with the goal of deploying this as a web-based
+tool, and ultimately renamed it to: USACE Coastal Engineering Toolkit (UCET).
+
+The RAD team focused on updating the Python codebase utilizing Python’s object-oriented programming and the newly
+developed HoloViz ecosystem. The team refactored the code to implement inheritance so the code is clean, readable, and
+scalable. The tools were expanded to a Graphical User Interface (GUI) so the implementation to a web-app would provide
```
Comment (KedoKudo) [inline papers/Catlett_Amanda/UCET.rst:58]: Could the authors also summaries the long-term maintainability of the modernized web app in the following areas:
- user documentation and user documentation update policy
- developer documentation
- overall code complexity

File: papers/Catlett_Amanda/UCET.rst [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+To make implementation of the GUI more seamless we refactored model variables to utilize the Param library. Param is a
+library that has the goal of simplifying the codebase by letting the programmer explicitly declare the types and values
+of parameters accepted by the code. Param can also be seamlessly used when implementing the GUI through Panel and
+HoloViews.
+
+Each UCET tool’s model class declares the input and output values used in the model as class variables. Each input and
+output variables are declared and given the following metadata features:
+
+- **default**:  each input variable is defined as a Param with a default value defined from the 1992 ACES user manual
+- **bounds**: each input variable is defined with range values defined in the 1992 ACES user manual
```
Comment (KedoKudo) [inline papers/Catlett_Amanda/UCET.rst:79]: Is the `1992 ACES user manual` distributed with the source code?  If not, I would recommend providing a reference or a link to this manual.

**Review by KedoKudo — CHANGES_REQUESTED** [review]
The overall writing is very good, and the description of the process of modernizing a 30+ years old application is both informative and educational.
In addition to the comments specific to certain paragraphs, I have a few more questions regarding the overall modernization process:

- Are there any efforts in adding unit tests for both the models as well as the UI aspect?  If so, what are the challenges and how did the development team manage to overcome them?

- What is the deployment strategy for this modernized web application?  Based on the content of the manuscript, it seems like the calculation and visualization are carried out by the same node.  A typical web app generally have the computing work done on the server side (dedicated worker nodes that can handle computational intensive workload) while the visualization is done at the client side (rendering reduced data and submit computing job to worker nodes).  Is this type of deployment already supported?  If so, how does it handle the authentication process, assuming that the computing resources requires some level of protection.

-  Are there any continuous integration (CI) and continuous deployment (CD) in place for the development of this web app?

**Comment by deniederhut** [comment]
@scoobies mark pending comment

File: papers/Catlett_Amanda/UCET.rst (hunk 1 above)
Comment (acatlett8) [reply papers/Catlett_Amanda/UCET.rst:58]: Thanks for your comments! 
- We have some tests that are testing the Python is getting the same answers as the old FORTRAN. The biggest problem is getting the data output from the FORTRAN.  But we do not have any for the UI aspect. I was unaware that was a thing. This is something we will look into for sure
- UCET right now does not have HEAVY calculations. so right now we don't think there is a need to have multiple servers. We do need to discuss how we are deploying this on a website and will add that information
- We will write about plans for the CICD in a future works selection. 
- I will also add a section about the long-term maintainability for user documentation, user documentation update policy, developer documentation, overall code complexity

File: papers/Catlett_Amanda/UCET.rst (hunk 2 above)
Comment (acatlett8) [reply papers/Catlett_Amanda/UCET.rst:79]: Yes! that was a failing on my part! I will put a reference to the manual!

**Comment by deniederhut** [comment]
Hi @acatlett8 ! How are the paper revisions coming along?

File: papers/Catlett_Amanda/UCET.rst [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+added to the toolkit. Param, Panel, and HoloViews work cohesively together to not only run the models but make a simple
+interface.
+
+Future work will involve expanding UCET to include current coastal engineering models and refining the code to make the
+application work more smoothly.
+
+References
+----------
+
+.. [Leenknecht] David A. Leenknecht, Andre Szuwalski, and Ann R. Sherlock. 1992. Automated Coastal Engineering System -Technical Reference. Technical report.
```
Comment (cbcunc) [inline papers/Catlett_Amanda/UCET.rst]: Need better citation. URL would help. One like https://usace.contentdm.oclc.org/digital/collection/p266001coll1/id/2321/ if no DOI available.

File: papers/Catlett_Amanda/UCET.rst [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+.. [panel] “Panel: A High-Level App and Dashboarding Solution for Python.” Panel 0.12.6 Documentation, Panel Contributors, 2019, https://panel.holoviz.org/.
+
+.. [holoviz] “High-Level Tools to Simplify Visualization in Python.” HoloViz 0.13.0 Documentation, HoloViz Authors, 2017, https://holoviz.org.
+
+.. [UG] David A. Leenknecht, et al. “Automated Tools for Coastal Engineering.” Journal of Coastal Research, vol. 11, no. 4, Coastal Education & Research Foundation, Inc., 1995, pp. 1108-24.
+
+.. [shankar] N.J. Shankar, M.P.R. Jayaratne, Wave run-up and overtopping on smooth and rough slopes of coastal structures, Ocean Engineering, Volume 30, Issue 2, 2003, Pages 221-238, ISSN 0029-8018,
```
Comment (cbcunc) [inline papers/Catlett_Amanda/UCET.rst]: DOI is https://doi.org/10.1016/S0029-8018(02)00016-1

**Review by cbcunc — CHANGES_REQUESTED** [review]

File: papers/Catlett_Amanda/UCET.rst (hunk 4 above)
Comment (cbcunc) [inline papers/Catlett_Amanda/UCET.rst]: Why is there a terminal comma?

**Review by cbcunc — CHANGES_REQUESTED** [review]

File: papers/Catlett_Amanda/UCET.rst (hunk 4 above)
Comment (acatlett8) [reply papers/Catlett_Amanda/UCET.rst]: added DOI to shankar

Comment (acatlett8) [reply papers/Catlett_Amanda/UCET.rst]: changed DOI

File: papers/Catlett_Amanda/UCET.rst (hunk 3 above)
Comment (acatlett8) [reply papers/Catlett_Amanda/UCET.rst]: added URL

File: papers/Catlett_Amanda/UCET.rst (hunk 1 above)
Comment (acatlett8) [reply papers/Catlett_Amanda/UCET.rst:58]: I pushed some changes. Another author will be coming along to add information about CICD. I tried to add most of what you commented on under a new section called current state.

**Comment by acatlett8** [comment]
I thought I pushed this Friday. but apparently that did not happen. Still trying to figure out Git.

**Comment by deniederhut** [comment]
No worries! Is this ready for re-review?

**Comment by acatlett8** [comment]
@deniederhut I think it is ready for re-review!! I have pushed some commits. Do I need to do anything else??

**Comment by cbcunc** [comment]
Reviewers, please give your final recommendations. The deadline for final review comments is tomorrow.

Authors, please respond to review comments with changes or comments which resolve the conversation. Final acceptance decisions are July 4.

File: papers/Catlett_Amanda/UCET.rst [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+
+Future work will involve expanding UCET to include current coastal engineering models, and completing the security
+vetting process to deploy to a publicly accessible website. We plan to incorporate an automated CI/CD to ensure smooth
+deployment of future versions. We also will continue to incorporate feedback from users and refine the code to ensure
+the application provides a quality user experience.
+
+References
+----------
+
+.. [Leenknecht] David A. Leenknecht, Andre Szuwalski, and Ann R. Sherlock. 1992. Automated Coastal Engineering System -Technical Reference. Technical report.
```
Comment (KedoKudo) [inline papers/Catlett_Amanda/UCET.rst]: @acatlett8 it seems like you most recent merge accidentally removed the URL that was added a few days ago. Please add it back.

**Review by KedoKudo — APPROVED** [review]
The URL for the first reference was removed during the last merge.  I would recommend adding it back if possible.

The rest of the paper looks good for this ongoing project.  I would recommend it for publishing.

File: papers/Catlett_Amanda/UCET.rst (hunk 5 above)
Comment (acatlett8) [reply papers/Catlett_Amanda/UCET.rst]: thanks! I think I added it back!!

**Review by KedoKudo — APPROVED** [review]
Thanks for adding the url back.

**Comment by deniederhut** [comment]
@cbcunc is this paper ready?

**Comment by deniederhut** [comment]
@cbcunc github shows that you are still requesting changes on this paper -- is it ready for publication?

**Review by cbcunc — APPROVED** [review]

**Comment by cbcunc** [comment]
@scoobies mark ready

**Pull request merged**